lib/puppet/functions/deprecation.rb

```diff
@@ -1,38 +1,38 @@
 # frozen_string_literal: true
 
-#   Function to print deprecation warnings, Logs a warning once for a given key.
-#
-# The uniqueness key - can appear once.
-# The msg is the message text including any positional information that is formatted by the
-# user/caller of the method.
-# It is affected by the puppet setting 'strict', which can be set to :error
-# (outputs as an error message), :off (no message / error is displayed) and :warning
-# (default, outputs a warning)  *Type*: String, String.
-#
+# @summary Function to print deprecation warnings, Logs a warning once for a given key.
 Puppet::Functions.create_function(:deprecation) do
   # @param key
-  # @param  message
-  # @return deprecated warnings
+  #   The uniqueness key.  This function logs once for any given key.
+  # @param message
+  #   Is the message text including any positional information that is formatted by the user/caller of the function.
+  # @param use_strict_setting
+  #   When `true`, (the default), the function is affected by the puppet setting 'strict', which can be set to :error
+  #   (outputs as an error message), :off (no message / error is displayed) and :warning
+  #   (default, outputs a warning).
   dispatch :deprecation do
     param 'String', :key
     param 'String', :message
+    optional_param 'Boolean', :use_strict_setting
   end
 
-  def deprecation(key, message)
+  def deprecation(key, message, use_strict_setting = true) # rubocop:disable Style/OptionalBooleanParameter
```

Comment:
can the default for `use_strict_setting` maybe come from an environment variable? At the moment CI jobs on Puppet 8 fail if they use the legacy functions and that's a nice way to identify them. It would be helpful to have an easy way to identify them in the future.

Reply:
Can we have both?

Reply:
Not sure.  We _could_ but I think I'd lean towards keeping it simple.  The existing environment variable code already looked hacky without adding yet more.  Functions whose behaviour changes based on something other than the function inputs seem like an antipattern to me.  Let's not make it worse?

> that's a nice way to identify them
But was it 'nice'?  The current behaviour has actually forced us to make all modules _depend_ on stdlib 9 in a very short space of time. That may have caused more harm than good.

Maybe a new puppet-lint plugin would be a better way to identify calls to non-namespaced functions?

Reply:
I'm pretty sure that the road to hell is paved with env vars...

Reply:
A new linter would certainly do the job.

Reply:
Not a blocker though, so maybe let this stew over the weekend and if there aren't any objections get it merged and a release out Monday??

Reply:
Sounds good to me.

Thanks for jumping on this!